web/pandas/pdeps/0013-logical-type-system.md

```diff
@@ -0,0 +1,161 @@
+
+  - Signed Integer
+  - Unsigned Integer
+  - Floating Point
+  - Fixed Point
+  - Boolean
+  - Date
+  - Datetime
+  - Duration
+  - Interval
+  - Period
+  - Binary
+  - String
+  - Dictionary
+  - List
+  - Struct
+
+To ensure we maintain all of the current functionality of our existing type system(s), a base type structure would need to look something like:
+
+```python
+class BaseType:
+
+    @property
+    def dtype_backend -> Literal["pandas", "numpy", "pyarrow"]:
+        """
+        Library is responsible for the array implementation
+        """
+        ...
+
+    @property
+    def physical_type:
+        """
+        How does the backend physically implement this logical type? i.e. our
+        logical type may be a "string" and we are using pyarrow underneath -
+        is it a pa.string(), pa.large_string(), pa.string_view() or something else?
+        """
+        ...
+
+    @property
+    def missing_value_marker -> pd.NA|np.nan:
+        """
+        Sentinel used to denote missing values
+        """
+        ...
+```
+
+The theory behind this PDEP is that most users _should not_ care about the physical type that is being used. But if the abstraction our logical type provides is too much, a user could at least inspect and potentially configure which physical type to use.
+
+With regards to how we may expose such types to end users, there are two currently recognized proposals. The first would use factory functions to create the logical type, i.e. something like:
+
+```python
+pd.Series(["foo", "bar", "baz"], dtype=pd.string()  # assumed common case
+pd.Series(["foo", "bar", "baz"], dtype=pd.string(missing_value_marker=np.nan)
+pd.Series(["foo", "bar", "baz"], dtype=pd.string(physical_type=pa.string_view())
+```
+
+Another approach would be to classes:
+
+```python
+pd.Series(["foo", "bar", "baz"], dtype=pd.StringDtype()
+pd.Series(["foo", "bar", "baz"], dtype=pd.StringDtype(missing_value_marker=np.nan)
+pd.Series(["foo", "bar", "baz"], dtype=pd.StringDtype(physical_type=pa.string_view())
+```
+Note that the class-based approach would reuse existing classes like ``pd.StringDtype`` but change their purpose, whereas the factory function would more explicitly be a new approach. This is an area that requires more discussion amongst the team.
+
+## String Type Arguments
+
+This PDEP proposes that we maintain only a small subset of string arguments that can be used to construct logical types. Those string arguments are:
+
+  - intXX
+  - uintXX
+  - floatXX
+  - string
+  - datetime64[unit]
+  - datetime64[unit, tz]
+
+However, new code should be encouraged to use the logical constructors outlined previously. Particularly for aggregate types, trying to encode all of the information into a string can become unwieldy. Instead, keyword argument use should be encouraged:
+
+```python
+pd.Series(dtype=pd.list(value_type=pd.string()))
+```
+
+## Bridging Type Systems
```

Comment:
Do you need to address if people can convert between two physical types that back the same logical type?

below, you discuss where operations between 2 objects of different physical types, but same logical type, cause a priority in the result. But what if users want to convert from one physical type to another?

Reply:
Great callout - I assume you have in mind something like adding a logical int + float.

I _think_ the same rules should still apply, i.e. if one is physically a pyarrow int and the other a numpy float you should end up with logical float type backed by a physical pyarrow float

Reply:
Or if I have a numpy backed string and I want to convert to a pyarrow backed string.

Reply:
Those would both be under the logical string type, so if you tried to concatenate them together you would get a pyarrow string per the rules outlined here. Does that make sense?

Reply:
No, I'm saying that let's say you know that you have a `Series` of logical type `string`, and you also know that it is backed by `numpy` string, and you want to convert that `Series` to have a physical type of `arrow`.  Don't you need a public API that lets you convert from one physical backing to another?

Reply:
Ah OK I see your point. I'm not sure - we already have keywords in the library like `dtype_backend` on the I/O side. My perhaps naive thought is we shouldn't offer anything new - a user could still stick with those keywords in the places where they exist if they needed finer grained control. 

Overall the goal of this PDEP is to move users away from trying to manage the physical type

Reply:
I also somewhat question the _why_ a user would want to do what you are asking about. I think that assumes they know something about the internals of pandas and feel like they can optimize it by swapping to certain types. But I think it is up to pandas to optimize that, not end users

Reply:
> No, I'm saying that let's say you know that you have a `Series` of logical type `string`, and you also know that it is backed by `numpy` string, and you want to convert that `Series` to have a physical type of `arrow`. Don't you need a public API that lets you convert from one physical backing to another?

Withe the current text of the PDEP, a user can do `ser.astype(pd.string(physical_type=pa.string())` (or with naming closer to what is currently being used, something like `ser.astype(pd.string(backend="pyarrow"))`). 

While the argument of the PDEP is that _most_ users won't have a need to do that, those explicit arguments to construct a logical dtype instance for a specific backend / physical type are still available for those cases where a user wants more tight control. 

So I think what you are asking is available?

Reply:
> I also somewhat question the why a user would want to do what you are asking about. I think that assumes they know something about the internals of pandas and feel like they can optimize it by changing to certain physical types. But I think it is up to pandas to optimize that, not end users

This is a good question!  The possible answers that come to mind are 1) because there are behavior differences, 2) because there are perf differences, 3) they shouldn't.

1) is the point I'm hung up on.

In case 3), that raises a question of why even _have_ multiple variants if there aren't meaningful differences between them.  That's pure maintenance burden.

Reply:
> This is a good question! The possible answers that come to mind are 1) because there are behavior differences, 2) because there are perf differences, 3) they shouldn't.

I was thinking it could be due to (2) - performance differences.  i.e., you read in some file that gives you one physical backend, and you now know you want to convert it to something else for performance.

> Withe the current text of the PDEP, a user can do ser.astype(pd.string(physical_type=pa.string()) (or with naming closer to what is currently being used, something like ser.astype(pd.string(backend="pyarrow"))).

Yes, I think `astype()` solves the issue I brought up.  But maybe it should be added to the PDEP.

Reply:
> In case 3), that raises a question of why even _have_ multiple variants if there aren't meaningful differences between them. That's pure maintenance burden.

A lot of this just has to do with the legacy of the pandas project. Of course, NumPy, our own extension types, and pyarrow would all look really similar if you only had non-nullable integral data. If we started as a project today would we choose to back both pyarrow and NumPy and have our own extension type system? I don't think so

But given that history the question becomes how do we best manage all three of these systems? The status quo is to make users responsible for knowing the intricacies of these systems and stating explicitly the types they want, while being sure to manage them over the full lifecycle. I think that's a lot to ask end users, especially people who just want to express something and have pandas solve it for them.

I still think the datetime issue is a great example. A lot of users familiar with databases may consider DATE / TIMESTAMP to be totally distinct types. What value add do we think there is to these not returning an equivalent type to our users:

```python
>>> pd.Series(["2024-01-01 00:00:00"], dtype="datetime64[us]").dt.date
0    2024-01-01
dtype: object
>>> pd.Series(["2024-01-01 00:00:00"], dtype=pd.ArrowDtype(pa.timestamp("us"))).dt.date
0    2024-01-01
dtype: date32[day][pyarrow]
```

If I understand @jbrockmendel your concern around _behavior differences_ I think this is what you are referring to. In the logical type world, we can abstract this and just work with datetimes / dates. 

Asking a user to manage all of this I think leads to a lot of confusion and causes divergence in how users are going to use the library

Reply:
Here's another good example - try reading a CSV file that looks like this:

```python
data = io.StringIO("""
timestamps,strings
2024-01-01 00:00:00,foo
2024-01-01 01:00:00,bar
2024-01-01 02:00:00,baz
""")
```

Let's say I wanted to read that file, set the timestamp as the index and derive a new date from it that truly is just the date itself (no timestamp). A default attempt:

```python
data.seek(0)
df = pd.read_csv(
    data,
    index_col="timestamps",
    parse_dates=["timestamps"]
).assign(date_col=lambda x: x.index.date)
df.dtypes
```

yields all column dtype return values:

```
strings     object
date_col    object
dtype: object
```

If you read that the pyarrow string implementation has better string handling, you may be tempted to try adding `dtype_backend="pyarrow"` to that read_csv call:

```python
data.seek(0)
df = pd.read_csv(
    data,
    dtype_backend="pyarrow",
    index_col="timestamps",
    parse_dates=["timestamps"],    
).assign(date_col=lambda x: x.index.date)
```

Unfortunately this fails:

```python
ValueError: not all elements from date_cols are numpy arrays
```

If you remove `parse_dates` and instead switch it to the pyarrow engine, you will get a different error:

```
In [82]: data.seek(0)
    ...: df = pd.read_csv(
    ...:     data,
    ...:     dtype_backend="pyarrow",
    ...:     engine="pyarrow",
    ...:     index_col="timestamps",
    ...: ).assign(date_col=lambda x: x.index.date)
AttributeError: 'Index' object has no attribute 'date'
```

You may _finally_ get close to the desired behavior if you forgo trying to set the pa.timestamp in the index:

```python
In [91]: data.seek(0)
    ...: df = pd.read_csv(
    ...:     data,
    ...:     dtype_backend="pyarrow",
    ...:     engine="pyarrow",
    ...: ).assign(date_col=lambda x: x["timestamps"].dt.date)
    ...: df.dtypes
Out[91]: 
timestamps    timestamp[s][pyarrow]
strings             string[pyarrow]
date_col       date32[day][pyarrow]
dtype: object
```

These different code paths are extremely confusing. Logical types would help us solve that because our behavior would simply focus on having timestamps, dates, and strings. The fact that NumPy only implements timestamps whereas pyarrow implements timestamps (in a very similar fashion to NumPy) but also implements dates should be of no concern to the user - we can just make this work

Reply:
> If I understand @jbrockmendel your concern around behavior differences I think this is what you are referring to. In the logical type world, we can abstract this and just work with datetimes / dates.

I feel like we are talking past each other.  You want datetime_series.dt.date to always return a pd.date() dtype.  Makes sense so far.  In #58315 you agreed that we don't want numpy_datetime_series.dt.date to return a pyarrow dtype.  So we would need to implement a non-pyarrow date dtype.  But [here](https://github.com/pandas-dev/pandas/pull/58455#discussion_r1582229324) you specifically said you don't want to implement a non-pyarrow date dtype.  I can't reconcile these three statements.

> Here's another good example - try reading a CSV file that looks like this:

My first thought here is that we shouldn't have both a dtype_backend and an engine keyword (xref #51846).

Second is that you want an Index with timestamp[pyarrow] dtype to be (or just behave like) a DatetimeIndex. 
 (presumably the same for a duration[pyarrow]->TimedeltaIndex, maybe others?).  As in #58307, I'd be fine with that.  I think that merits its own discussion, since the implementation would be a pretty big hassle, perf questions to address, and the possibility of going the other direction and removing DatetimeIndex altogether (xref #43002).

Big picture: I'm less convinced that I know what this PDEP is focused on than i originally was.  I thought it was about how users specify `dtype=...`, but maybe that's only part of it?  And parts of the conversation seem to hinge on resolution to nan-vs-NA (and NaT, and None, and...) or on getting rid of support for numpy dtypes altogether, which I really hope can be orthogonal to "how do users specify dtypes"

Reply:
> In #58315 you agreed that we don't want numpy_datetime_series.dt.date to return a pyarrow dtype. So we would need to implement a non-pyarrow date dtype. But [here](https://github.com/pandas-dev/pandas/pull/58455#discussion_r1582229324) you specifically said you don't want to implement a non-pyarrow date dtype

Yea I don't think `numpy_datetime_series.dt.date` should return a pyarrow type. But I do think a logical `pd.datetime` which happens to be backed by NumPy is certainly free to return a `pd.date` logical type that can be backed by pyarrow. The end user is not burdened with understanding the different type systems in this case - we manage that as an implementation detail. Does that help clarify?

Reply:
So you _would_ want to implement a new date dtype with numpy semantics?  Just with it being backed by pyarrow?

Reply:
> So you _would_ want to implement a new date dtype with numpy semantics?

No - there would be no need at all for that. In this particular case, if the logical requirement was to go from a datetime -> date, and physically our array happened to be backed by a NumPy array, we internally would go NumPy datetime -> pa.timestamp -> pa.date; if the array was already backed by pyarrow we could skip the first step

> This can get tricky if/when PDEP-10 is reverted.

I _think_ even if PDEP-10 got reverted it would limit the logical types we offer (we'd essentially scrap date, list, dictionary, struct, etc...anything where only pyarrow implements something) until pyarrow becomes a hard dependency again

Reply:
> No - there would be no need at all for that [...]

We're definitely talking past each other, because AFAICT the suggestion here directly contradicts discussion in #58315.  Let's put a pin in this until the call on Wednesday and see if we can sort it out then.

Reply:
Summarizing this part of the discussion from the call: the communication failure was in 1) what "mixed semantics" I was concerned about and 2) what Will meant by not needing to implement anything new.

Consider

```
dti = pd.date_range("2016-01-01", periods=3)
ser = pd.Series(dti)
ser.iloc[1] = pd.NaT
df = ser.to_frame("datetime")
df["date"] = df["datetime"].dt.date
```

Now suppose that df["date"] is backed (perhaps indirectly) by a pyarrow array.  When I say the columns have "mixed semantics", I am referring to the propagation of nulls in comparisons

```
mask1 = df["date"] == dti[0].date()
mask2 = df["date"] != dti[0].date()
```

In this scenario `mask1` and `mask2` are both nullable and their second entry is pd.NA.  Using these masks for indexing gives surprising results (raising ATM, different if the Ice Cream Agreement goes into effect) to users expecting numpy semantics.  <b>A user who has not opted in to null propagation has some columns that behave this way and some that do not.</b>  Avoiding that scenario is why we implemented "string[pyarrow_numpy]" to have arrow storage with numpy semantics.

OK, all of that is what I had failed to communicate.  What Will was saying when he said (paraphrasing) "we wouldn't need to implement another dtype", that didn't exclude a thin wrapper to retain numpy propagation semantics with pyarrow storage.

@WillAyd can you confirm that the above is a valid summary of what we got sorted out at the end of the call?

Reply:
Yea that's correct and a nice summary. This is why the BaseType in the PDEP has the `missing_value_marker` / `na_value` - it explicitly decouples the NA handling from the physical type. It specifically allows the logical type to cover the functionality of `string[pyarrow_numpy]` as you described above, and as we talked about on the call theoretically also allows you to have a logical date type which uses pyarrow behind the scenes but can use `pd.NaT` as a `na_value`.

I think this PDEP could help achieve more consistent NA handling over time since it decouples the user expectations from a "backend", but actually propogating one NA marker consistently is left to a separate PDEP